crates/lsp/src/main_loop.rs

```diff
@@ -288,28 +245,30 @@ impl GlobalState {
     /// This takes ownership of all global state and handles one by one LSP
     /// requests, notifications, and other internal events.
     async fn main_loop(mut self) {
-        // Spawn latency-sensitive auxiliary loop. Must be first to initialise
-        // global transmission channel.
+        // Spawn latency-sensitive auxiliary and log threads.
         let mut set = tokio::task::JoinSet::<()>::new();
 
-        // Move the `auxiliary_state` owned by the global state to its own thread.
-        // Unwrap: `start()` should only ever be called once.
+        // Take ownership over `log_state` and start the log thread.
+        // Unwrap: `start()` should only be called once.
+        let log_state = self.log_state.take().unwrap();
+        set.spawn(async move { log_state.start().await });
+
+        // Take ownership over `auxiliary_state` and start the auxiliary thread.
+        // Unwrap: `start()` should only be called once.
         let auxiliary_state = self.auxiliary_state.take().unwrap();
         set.spawn(async move { auxiliary_state.start().await });
 
         loop {
             let event = self.next_event().await;
             match self.handle_event(event).await {
-                Err(err) => self.log_error(format!("Failure while handling event:\n{err:?}")),
+                Err(err) => tracing::error!("Failure while handling event:\n{err:?}"),
                 Ok(LoopControl::Shutdown) => break,
                 _ => {}
             }
         }
 
-        log::trace!("Main loop closed. Shutting down auxiliary loop.");
+        tracing::trace!("Main loop closed. Shutting down auxiliary and log loop.");
         set.shutdown().await;
-
-        log::trace!("Main loop exited.");
     }
```

Comment:
Remember that after the log loop is shutdown, we won't be able to see anymore messages. I've removed a few messages that would otherwise occur too soon or too late because of this to prevent confusion.